File: talazo/ai_recommendations.py
# talazo/ml/ai_recommendations.py
import requests
import os
import json
import logging
from datetime import datetime, timedelta
from functools import lru_cache
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

logger = logging.getLogger(__name__)

class AIRecommendationSystem:
    """
    AI-powered recommendation system for agricultural advice
    Uses the Groq API to generate contextual recommendations based on soil data
    """
    
    def __init__(self, cache_expiry=3600):
        """
        Initialize the AI Recommendation System
        
        Args:
            cache_expiry (int, optional): Cache expiry time in seconds. Defaults to 3600 (1 hour).
        """
        self.api_key = os.environ.get('GROQ_API_KEY')
        self.api_url = 'https://api.groq.com/openai/v1/chat/completions'
        self.model = 'llama3-8b-8192'  # or 'mixtral-8x7b-32768' depending on needs
        self.cache_expiry = cache_expiry
        
        # Initialize request session for connection pooling
        self.session = requests.Session()
        
        if not self.api_key:
            logger.warning("GROQ_API_KEY not configured. AI recommendations will use fallback system.")
    
    def generate_recommendations(self, soil_data, farmer_info=None, region=None, crop=None):
        """
        Generate AI-driven recommendations based on soil data
        
        Args:
            soil_data (dict): Dictionary of soil parameters
            farmer_info (dict, optional): Farmer details for contextual recommendations
            region (str, optional): Geographic region in Zimbabwe
            crop (str, optional): Current or planned crop
            
        Returns:
            dict: Dictionary containing recommendations and metadata
        """
        # Check if API key is available
        if not self.api_key:
            logger.warning("Using fallback recommendations due to missing API key")
            return {
                'status': 'fallback',
                'recommendations': self._fallback_recommendations(soil_data, crop),
                'metadata': {
                    'source': 'fallback',
                    'timestamp': datetime.utcnow().isoformat()
                }
            }
        
        # Try to get cached recommendations if parameters are similar
        cache_key = self._generate_cache_key(soil_data, region, crop)
        cached_result = self._get_cached_recommendation(cache_key)
        if cached_result:
            logger.info("Returning cached recommendation")
            return cached_result
        
        # Create prompt with soil data
        prompt = self._create_prompt(soil_data, farmer_info, region, crop)
        
        try:
            # Call Groq LLM API
            headers = {
                'Authorization': f'Bearer {self.api_key}',
                'Content-Type': 'application/json'
            }
            
            data = {
                'model': self.model,
                'messages': [
                    {'role': 'system', 'content': 'You are an expert agricultural advisor specializing in Zimbabwean farming conditions. Provide practical, actionable recommendations based on soil health data that are appropriate for small-scale farmers in Zimbabwe.'},
                    {'role': 'user', 'content': prompt}
                ],
                'temperature': 0.2,  # Lower temperature for more consistent outputs
                'max_tokens': 1024
            }
            
            logger.debug(f"Sending request to Groq API for soil data: {json.dumps(soil_data)}")
            response = self.session.post(
                self.api_url, 
                headers=headers, 
                json=data, 
                timeout=30  # Add timeout to prevent hanging
            )
            
            if response.status_code == 200:
                response_data = response.json()
                recommendations_text = response_data['choices'][0]['message']['content']
                
                # Parse the recommendations
                parsed_recommendations = self._parse_recommendations(recommendations_text)
                
                result = {
                    'status': 'success',
                    'recommendations': parsed_recommendations,
                    'raw_response': recommendations_text,
                    'metadata': {
                        'source': 'groq',
                        'model': self.model,
                        'timestamp': datetime.utcnow().isoformat()
                    }
                }
                
                # Cache the result
                self._cache_recommendation(cache_key, result)
                return result
            else:
                logger.error(f"Groq API request failed with status {response.status_code}: {response.text}")
                # API call failed, use fallback
                return {
                    'status': 'error',
                    'error': f'API call failed with status {response.status_code}',
                    'recommendations': self._fallback_recommendations(soil_data, crop),
                    'metadata': {
                        'source': 'fallback',
                        'timestamp': datetime.utcnow().isoformat()
                    }
                }
                
        except Exception as e:
            logger.exception(f"Error in AI recommendation generation: {str(e)}")
            # Error occurred, use fallback
            return {
                'status': 'error',
                'error': str(e),
                'recommendations': self._fallback_recommendations(soil_data, crop),
                'metadata': {
                    'source': 'fallback',
                    'timestamp': datetime.utcnow().isoformat()
                }
            }
    
    def _create_prompt(self, soil_data, farmer_info=None, region=None, crop=None):
        """
        Create a detailed prompt for the LLM based on soil data
        
        Args:
            soil_data (dict): Dictionary of soil parameters
            farmer_info (dict, optional): Farmer details for contextual recommendations
            region (str, optional): Geographic region in Zimbabwe
            crop (str, optional): Current or planned crop
            
        Returns:
            str: Formatted prompt for the LLM
        """
        # Format soil data
        soil_info = '\n'.join([f'- {k.replace("_", " ").title()}: {v}' for k, v in soil_data.items()])
        
        # Get current season
        current_month = datetime.now().month
        season = self._determine_season(current_month)
        
        # Build the prompt
        prompt = f"""
        Analyze the following soil health data from a farm in {region or 'Zimbabwe'} and provide detailed recommendations:

        SOIL DATA:
        {soil_info}

        CONTEXT:
        - Current season: {season}
        - Region: {region or 'Zimbabwe'}
        """
        
        if crop:
            prompt += f"\n- Current/planned crop: {crop}"
        
        if farmer_info:
            farmer_context = '\n'.join([f'- {k.replace("_", " ").title()}: {v}' for k, v in farmer_info.items()])
            prompt += f"\n\nFARMER INFORMATION:\n{farmer_context}"
        
        prompt += """
        
        Please provide the following in a structured format:
        
        1. SOIL HEALTH ANALYSIS: Explain what these soil readings indicate about soil health.
        
        2. PRIORITY RECOMMENDATIONS: List 3-5 specific actions the farmer should take to improve soil health, in priority order. For each recommendation, include:
           - ACTION: The specific action to take
           - REASON: Why it's needed based on the soil data
           - COST: Estimated cost (low/medium/high)
           - TIMEFRAME: Expected timeframe for results
           - LOCAL CONSIDERATIONS: Any Zimbabwe-specific factors
        
        3. CROP RECOMMENDATIONS: Suggest 2-3 crops that would be suitable for these soil conditions, explaining why each is appropriate.
        
        4. SUSTAINABLE PRACTICES: Recommend 1-2 sustainable farming practices that would improve long-term soil health.
        
        Format each recommendation as a clear section with the headings above. Be practical, specific, and consider the constraints of small-scale farmers in Zimbabwe.
        """
        
        return prompt
    
    def _parse_recommendations(self, recommendations_text):
        """
        Parse the LLM response into structured recommendations
        
        Args:
            recommendations_text (str): Raw text response from the LLM
            
        Returns:
            list: List of structured recommendation dictionaries
        """
        # Split the text into sections
        sections = []
        current_section = ""
        current_heading = None
        
        for line in recommendations_text.split('\n'):
            line = line.strip()
            if not line:
                continue
                
            # Check if this is a new section heading
            if line.upper() == "SOIL HEALTH ANALYSIS:" or "SOIL HEALTH ANALYSIS" in line.upper():
                if current_heading:
                    sections.append({"type": current_heading, "content": current_section.strip()})
                current_heading = "soil_analysis"
                current_section = ""
            elif line.upper() == "PRIORITY RECOMMENDATIONS:" or "PRIORITY RECOMMENDATIONS" in line.upper():
                if current_heading:
                    sections.append({"type": current_heading, "content": current_section.strip()})
                current_heading = "recommendations"
                current_section = ""
            elif line.upper() == "CROP RECOMMENDATIONS:" or "CROP RECOMMENDATIONS" in line.upper():
                if current_heading:
                    sections.append({"type": current_heading, "content": current_section.strip()})
                current_heading = "crops"
                current_section = ""
            elif line.upper() == "SUSTAINABLE PRACTICES:" or "SUSTAINABLE PRACTICES" in line.upper():
                if current_heading:
                    sections.append({"type": current_heading, "content": current_section.strip()})
                current_heading = "sustainable_practices"
                current_section = ""
            else:
                current_section += line + "\n"
        
        # Add the last section
        if current_heading and current_section:
            sections.append({"type": current_heading, "content": current_section.strip()})
        
        # Process the recommendations section into structured format
        structured_recommendations = []
        
        for section in sections:
            if section["type"] == "recommendations":
                recommendations_text = section["content"]
                recommendation_blocks = recommendations_text.split("\n\n")
                
                for block in recommendation_blocks:
                    if not block.strip():
                        continue
                        
                    rec = {"type": "recommendation"}
                    
                    # Extract title
                    if ":" in block.split("\n")[0]:
                        title_parts = block.split("\n")[0].split(":", 1)
                        rec["title"] = title_parts[1].strip() if len(title_parts) > 1 else title_parts[0].strip()
                    else:
                        rec["title"] = block.split("\n")[0].strip()
                    
                    # Extract other fields
                    for line in block.split("\n"):
                        line = line.strip()
                        if not line or ":" not in line:
                            continue
                            
                        key, value = line.split(":", 1)
                        key = key.strip().upper()
                        
                        if "ACTION" in key:
                            rec["action"] = value.strip()
                        elif "REASON" in key or "WHY" in key:
                            rec["reason"] = value.strip()
                        elif "COST" in key:
                            rec["cost_estimate"] = value.strip()
                        elif "TIMEFRAME" in key or "TIME" in key:
                            rec["timeframe"] = value.strip()
                        elif "LOCAL" in key or "CONSIDERATION" in key:
                            rec["local_context"] = value.strip()
                    
                    if len(rec) > 1:  # Only add if we have more than just the title
                        structured_recommendations.append(rec)
        
        # If we couldn't parse properly, create a simple recommendation
        if not structured_recommendations and recommendations_text:
            structured_recommendations.append({
                "title": "Soil Improvement",
                "action": "Refer to the detailed text analysis",
                "content": recommendations_text
            })
        
        return structured_recommendations
    
    def _determine_season(self, month):
        """
        Determine current agricultural season in Zimbabwe based on month
        
        Args:
            month (int): Month as integer (1-12)
            
        Returns:
            str: Current agricultural season
        """
        if month in [11, 12, 1, 2, 3]:
            return "Rainy Season (Summer)"
        elif month in [5, 6, 7, 8]:
            return "Dry Season (Winter)"
        elif month in [4, 9, 10]:
            return "Transition Season"
    
    def _fallback_recommendations(self, soil_data, crop=None):
        """
        Provide fallback recommendations if API call fails
        
        Args:
            soil_data (dict): Dictionary of soil parameters
            crop (str, optional): Current or planned crop
            
        Returns:
            list: List of recommendation dictionaries
        """
        recommendations = []
        
        # Check pH level
        try:
            ph = float(soil_data.get('ph_level', 0))
            if ph < 6.0:
                recommendations.append({
                    'title': 'Correct Soil pH',
                    'action': 'Apply agricultural lime at 2-4 tons per hectare',
                    'reason': 'Soil is too acidic, which limits nutrient availability',
                    'cost_estimate': 'Medium',
                    'timeframe': '3-6 months',
                    'local_context': 'Locally available agricultural lime can be sourced from most agricultural supply stores in Zimbabwe'
                })
            elif ph > 7.0:
                recommendations.append({
                    'title': 'Reduce Soil pH',
                    'action': 'Apply organic matter such as compost or manure',
                    'reason': 'Soil is too alkaline, which can limit certain nutrient uptake',
                    'cost_estimate': 'Low to Medium',
                    'timeframe': '3-6 months',
                    'local_context': 'Manure from local livestock can be an affordable option'
                })
        except (ValueError, TypeError):
            pass
        
        # Check nitrogen level
        try:
            nitrogen = float(soil_data.get('nitrogen_level', 0))
            if nitrogen < 20:
                recommendations.append({
                    'title': 'Increase Nitrogen Levels',
                    'action': 'Apply nitrogen fertilizer (Ammonium Nitrate) at 100-150 kg/ha',
                    'reason': 'Nitrogen deficiency will limit plant growth and yield',
                    'cost_estimate': 'Medium',
                    'timeframe': '2-4 weeks',
                    'local_context': 'Consider split application during the growing season for better efficiency in Zimbabwe\'s climate'
                })
        except (ValueError, TypeError):
            pass
        
        # Check organic matter
        try:
            organic_matter = float(soil_data.get('organic_matter', 0))
            if organic_matter < 3.0:
                recommendations.append({
                    'title': 'Increase Organic Matter',
                    'action': 'Apply compost or incorporate crop residues into the soil',
                    'reason': 'Low organic matter reduces soil structure, water retention, and nutrient availability',
                    'cost_estimate': 'Low',
                    'timeframe': '6-12 months',
                    'local_context': 'Conservation agriculture techniques are being promoted in Zimbabwe and can help build organic matter over time'
                })
        except (ValueError, TypeError):
            pass
        
        # Check moisture content
        try:
            moisture = float(soil_data.get('moisture_content', 0))
            if moisture < 20:
                recommendations.append({
                    'title': 'Improve Water Management',
                    'action': 'Apply mulch and implement water conservation practices',
                    'reason': 'Low soil moisture will stress plants and reduce yields',
                    'cost_estimate': 'Low to Medium',
                    'timeframe': 'Immediate benefits',
                    'local_context': 'Given Zimbabwe\'s drought-prone regions, water conservation is critical'
                })
        except (ValueError, TypeError):
            pass
        
        # Add crop-specific recommendation if crop is provided
        if crop:
            crop = crop.lower()
            if crop == 'maize':
                recommendations.append({
                    'title': 'Maize-Specific Management',
                    'action': 'Ensure proper spacing at 75cm between rows and 25cm between plants',
                    'reason': 'Proper spacing optimizes resource use and yield potential',
                    'cost_estimate': 'Low',
                    'timeframe': 'Planting time',
                    'local_context': 'Maize is a staple crop in Zimbabwe, and proper spacing is essential for good yields'
                })
            elif crop == 'groundnuts' or crop == 'peanuts':
                recommendations.append({
                    'title': 'Groundnut Management',
                    'action': 'Ensure calcium levels are adequate with gypsum application if needed',
                    'reason': 'Calcium is essential for pod development in groundnuts',
                    'cost_estimate': 'Medium',
                    'timeframe': 'Apply at flowering',
                    'local_context': 'Groundnuts are an important cash and food security crop in Zimbabwe'
                })
        
        return recommendations
    
    def _generate_cache_key(self, soil_data, region=None, crop=None):
        """
        Generate a cache key based on input parameters
        
        Args:
            soil_data (dict): Dictionary of soil parameters
            region (str, optional): Geographic region
            crop (str, optional): Current/planned crop
            
        Returns:
            str: Cache key
        """
        # Round numerical values to reduce cache variations
        rounded_data = {}
        for key, value in soil_data.items():
            try:
                if isinstance(value, (int, float)):
                    rounded_data[key] = round(float(value), 1)
                else:
                    rounded_data[key] = value
            except (ValueError, TypeError):
                rounded_data[key] = value
        
        # Create a string representation for hashing
        key_parts = [
            json.dumps(rounded_data, sort_keys=True),
            region or '',
            crop or ''
        ]
        
        return '_'.join(key_parts)
    
    # Use LRU cache for storing recommendations
    @lru_cache(maxsize=100)
    def _get_cached_recommendation(self, cache_key):
        """Placeholder for retrieving cached recommendations"""
        # In a real implementation, this would check a persistent cache
        return None
    
    def _cache_recommendation(self, cache_key, result):
        """Placeholder for storing recommendations in cache"""
        # In a real implementation, this would store in a persistent cache
        pass

File: talazo/static/js/ai_recommendations.js
// static/js/ai_recommendations.js - Enhanced version
class AIRecommendationManager {
  constructor() {
      this.apiEndpoint = '/api/soil/ai-recommendations';
      this.loadingElement = document.getElementById('ai-recommendations-loading');
      this.contentElement = document.getElementById('ai-recommendations-content');
      this.generateButton = document.getElementById('generate-recommendations-btn');
      this.recommendations = [];
      this.lastGeneratedTime = null;
      
      this.init();
  }
  
  init() {
      console.log('Initializing AI Recommendation Manager...');
      
      if (this.generateButton) {
          this.generateButton.addEventListener('click', () => this.generateRecommendations());
      }
      
      // Add print and share functionality
      this.setupActionButtons();
      
      // Check if we should auto-generate recommendations
      if (this.contentElement && this.contentElement.children.length === 0) {
          // Auto-generate after a brief delay to let the dashboard load
          setTimeout(() => {
              // Only auto-generate if we haven't already done so
              if (!this.lastGeneratedTime) {
                  this.generateRecommendations();
              }
          }, 2000);
      }
  }
  
  setupActionButtons() {
      // Setup action buttons in the card header if they exist
      const printButton = document.querySelector('.ai-recommendations .fa-print');
      if (printButton && printButton.parentElement) {
          printButton.parentElement.addEventListener('click', () => this.printRecommendations());
      }
      
      const shareButton = document.querySelector('.ai-recommendations .fa-share-alt');
      if (shareButton && shareButton.parentElement) {
          shareButton.parentElement.addEventListener('click', () => this.shareRecommendations());
      }
  }
  
  async generateRecommendations() {
      // Show loading state
      this.showLoading();
      
      try {
          // Get current soil data from dashboard if available
          let soilData = {};
          let region = 'Zimbabwe';
          let crop = 'Maize';
          
          // Try to get soil data from dashboard
          if (window.dashboardManager && window.dashboardManager.currentSoilData) {
              soilData = window.dashboardManager.currentSoilData;
          } else {
              // If dashboard data is not available, create sample data
              soilData = this.createSampleSoilData();
          }
          
          // Get farmer selection if available
          const farmerSelector = document.getElementById('farmer-selector');
          if (farmerSelector && farmerSelector.selectedOptions && farmerSelector.selectedOptions.length > 0) {
              const farmerText = farmerSelector.selectedOptions[0].text;
              if (farmerText.includes('(')) {
                  region = farmerText.split('(')[1].replace(')', '').trim();
              }
          }
          
          // Try to get API data
          let data;
          try {
              // Make API request
              const response = await fetch(this.apiEndpoint, {
                  method: 'POST',
                  headers: {
                      'Content-Type': 'application/json'
                  },
                  body: JSON.stringify({
                      soil_data: soilData,
                      region: region,
                      crop: crop,
                      farmer_info: {
                          primary_crop: crop,
                          farm_size: 2.5
                      }
                  })
              });
              
              if (!response.ok) {
                  throw new Error(`API request failed with status ${response.status}`);
              }
              
              data = await response.json();
              
          } catch (error) {
              console.warn('API request failed, using fallback recommendations', error);
              // Use fallback recommendations
              data = this.getFallbackRecommendations(soilData, region, crop);
          }
          
          this.recommendations = data.recommendations || [];
          this.lastGeneratedTime = new Date();
          this.displayRecommendations(data);
          
      } catch (error) {
          console.error('Error generating AI recommendations:', error);
          this.showError('Failed to generate recommendations. Using fallback recommendations instead.');
          
          // Use fallback in case of errors
          const data = this.getFallbackRecommendations(this.createSampleSoilData(), 'Zimbabwe', 'Maize');
          this.recommendations = data.recommendations || [];
          this.displayRecommendations(data);
      } finally {
          this.hideLoading();
      }
  }
  
  createSampleSoilData() {
      // Create sample soil data if none is available
      return {
          ph_level: 6.2,
          nitrogen_level: 25,
          phosphorus_level: 18,
          potassium_level: 180,
          organic_matter: 3.5,
          cation_exchange_capacity: 14,
          moisture_content: 22
      };
  }
  
  getFallbackRecommendations(soilData, region, crop) {
      // Generate reasonable recommendations based on soil data
      const recommendations = [];
      
      // Check pH level
      if (soilData.ph_level < 6.0) {
          recommendations.push({
              title: 'Correct Soil pH',
              action: `Apply agricultural lime at ${Math.round((6.5 - soilData.ph_level) * 2)} tons per hectare`,
              reason: 'Soil is too acidic, which limits nutrient availability',
              cost_estimate: 'Medium',
              timeframe: '3-6 months',
              local_context: 'Locally available agricultural lime can be sourced from most agricultural supply stores in Zimbabwe'
          });
      } else if (soilData.ph_level > 7.0) {
          recommendations.push({
              title: 'Reduce Soil pH',
              action: 'Apply organic matter such as compost or manure',
              reason: 'Soil is too alkaline, which can limit certain nutrient uptake',
              cost_estimate: 'Low to Medium',
              timeframe: '3-6 months',
              local_context: 'Manure from local livestock can be an affordable option'
          });
      }
      
      // Check nitrogen level
      if (soilData.nitrogen_level < 20) {
          recommendations.push({
              title: 'Increase Nitrogen Levels',
              action: `Apply nitrogen fertilizer (Ammonium Nitrate) at ${Math.round(150 - soilData.nitrogen_level * 5)} kg/ha`,
              reason: 'Nitrogen deficiency will limit plant growth and yield',
              cost_estimate: 'Medium',
              timeframe: '2-4 weeks',
              local_context: 'Consider split application during the growing season for better efficiency in Zimbabwe\'s climate'
          });
      }
      
      // Check organic matter
      if (soilData.organic_matter < 3.0) {
          recommendations.push({
              title: 'Increase Organic Matter',
              action: 'Apply compost or incorporate crop residues into the soil',
              reason: 'Low organic matter reduces soil structure, water retention, and nutrient availability',
              cost_estimate: 'Low',
              timeframe: '6-12 months for full benefits',
              local_context: 'Conservation agriculture techniques are promoted in Zimbabwe to build organic matter over time'
          });
      }
      
      // Check moisture content
      if (soilData.moisture_content < 20) {
          recommendations.push({
              title: 'Improve Water Management',
              action: 'Apply mulch and implement water conservation practices like tied ridges or basins',
              reason: 'Low soil moisture will stress plants and reduce yields',
              cost_estimate: 'Low to Medium',
              timeframe: 'Immediate benefits',
              local_context: 'Critical for rainfed agriculture in Zimbabwe\'s drought-prone regions'
          });
      }
      
      return {
          status: 'success',
          recommendations: recommendations,
          metadata: {
              source: 'fallback',
              request_context: {
                  region: region,
                  crop: crop
              },
              timestamp: new Date().toISOString()
          }
      };
  }
  
  displayRecommendations(data) {
      if (!data.recommendations || data.status === 'error') {
          this.showError(data.error || 'An error occurred generating recommendations');
          return;
      }
      
      // Clear previous content
      this.contentElement.innerHTML = '';
      
      // Create recommendations UI
      const recommendations = data.recommendations || [];
      
      if (recommendations.length === 0) {
          this.contentElement.innerHTML = `
              <div class="empty-state">
                  <i class="fas fa-robot"></i>
                  <p>No specific recommendations available for the current soil conditions.</p>
              </div>
          `;
          this.contentElement.classList.remove('hidden');
          return;
      }
      
      // Add header with metadata
      const header = document.createElement('div');
      header.className = 'ai-recommendations-header';
      header.innerHTML = `
          <div class="metadata">
              <div class="meta-item">
                  <span class="label">Generated:</span>
                  <span class="value">${new Date().toLocaleString()}</span>
              </div>
              <div class="meta-item">
                  <span class="label">Region:</span>
                  <span class="value">${data.metadata?.request_context?.region || 'Zimbabwe'}</span>
              </div>
              <div class="meta-item">
                  <span class="label">Crop:</span>
                  <span class="value">${data.metadata?.request_context?.crop || 'General'}</span>
              </div>
          </div>
      `;
      this.contentElement.appendChild(header);
      
      // Create recommendation cards
      const recList = document.createElement('div');
      recList.className = 'recommendation-list';
      
      recommendations.forEach((rec, index) => {
          const card = this.createRecommendationCard(rec, index + 1);
          recList.appendChild(card);
      });
      
      this.contentElement.appendChild(recList);
      
      // Show the content
      this.contentElement.classList.remove('hidden');
  }
  
  createRecommendationCard(recommendation, index) {
      const card = document.createElement('div');
      card.className = 'recommendation-card';
      card.setAttribute('data-rec-id', `ai-rec-${index}`);
      
      // Extract recommendation details
      const title = recommendation.title || 'Soil Improvement Recommendation';
      const action = recommendation.action || '';
      const reason = recommendation.reason || '';
      const costEstimate = recommendation.cost_estimate || 'Unknown';
      const timeframe = recommendation.timeframe || '';
      const localContext = recommendation.local_context || '';
      
      // Build card content
      let cardContent = `
          <div class="recommendation-header">
              <div class="recommendation-number">${index}</div>
              <h3>${title}</h3>
          </div>
          <div class="recommendation-body">
      `;
      
      // Add recommendation details
      if (action) {
          cardContent += `<p class="recommendation-action"><strong>Action:</strong> ${action}</p>`;
      }
      
      if (reason) {
          cardContent += `<p><strong>Why:</strong> ${reason}</p>`;
      }
      
      if (costEstimate) {
          cardContent += `<p><strong>Estimated Cost:</strong> <span class="cost-tag ${this.getCostClass(costEstimate)}">${costEstimate}</span></p>`;
      }
      
      if (timeframe) {
          cardContent += `<p><strong>Timeframe:</strong> ${timeframe}</p>`;
      }
      
      if (localContext) {
          cardContent += `<div class="local-context"><strong>Local Considerations:</strong> ${localContext}</div>`;
      }
      
      cardContent += `
          </div>
          <div class="recommendation-footer">
              <button class="btn btn-sm implement-btn">Implement</button>
              <button class="btn btn-sm btn-outline save-btn">Save for Later</button>
          </div>
      `;
      
      card.innerHTML = cardContent;
      
      // Add event listeners
      card.querySelector('.implement-btn').addEventListener('click', () => this.implementRecommendation(index));
      card.querySelector('.save-btn').addEventListener('click', () => this.saveRecommendation(index));
      
      return card;
  }
  
  implementRecommendation(index) {
      if (index >= this.recommendations.length) return;
      
      const recommendation = this.recommendations[index];
      console.log('Implementing recommendation:', recommendation);
      
      // Show confirmation notification
      if (window.dashboardManager) {
          window.dashboardManager.showNotification(`Implementing "${recommendation.title || 'recommendation'}"`, 'success');
      }
      
      // Update card to show implemented status
      const card = document.querySelector(`[data-rec-id="ai-rec-${index}"]`);
      if (card) {
          card.classList.add('implemented');
          card.querySelector('.recommendation-footer').innerHTML = `
              <div class="implemented-status">
                  <i class="fas fa-check-circle"></i> Implementation in progress
              </div>
          `;
      }
  }
  
  saveRecommendation(index) {
      if (index >= this.recommendations.length) return;
      
      const recommendation = this.recommendations[index];
      console.log('Saving recommendation:', recommendation);
      
      // Show confirmation notification
      if (window.dashboardManager) {
          window.dashboardManager.showNotification(`Saved "${recommendation.title || 'recommendation'}" for later`, 'default');
      }
      
      // Update button to show saved status
      const card = document.querySelector(`[data-rec-id="ai-rec-${index}"]`);
      if (card) {
          const saveBtn = card.querySelector('.save-btn');
          if (saveBtn) {
              saveBtn.innerHTML = '<i class="fas fa-check"></i> Saved';
              saveBtn.classList.add('saved');
              saveBtn.disabled = true;
          }
      }
  }
  
  printRecommendations() {
      if (this.recommendations.length === 0) {
          if (window.dashboardManager) {
              window.dashboardManager.showNotification('No recommendations to print', 'warning');
          }
          return;
      }
      
      // Create a printable version
      const printWindow = window.open('', '_blank', 'width=800,height=600');
      if (!printWindow) {
          alert('Please allow pop-ups to print recommendations');
          return;
      }
      
      const html = `
          <!DOCTYPE html>
          <html>
          <head>
              <title>Talazo AgriFinance - AI Recommendations</title>
              <style>
                  body { font-family: Arial, sans-serif; margin: 20px; color: #333; }
                  h1 { color: #1E8449; }
                  .recommendation { margin-bottom: 20px; padding: 15px; border: 1px solid #ddd; page-break-inside: avoid; }
                  .recommendation h3 { margin-top: 0; color: #1E8449; }
                  .metadata { margin-bottom: 20px; color: #666; font-size: 0.9em; }
                  .footer { margin-top: 30px; font-size: 0.8em; color: #666; text-align: center; }
              </style>
          </head>
          <body>
              <h1>Talazo AgriFinance - AI Recommendations</h1>
              <div class="metadata">
                  <p>Generated: ${new Date().toLocaleString()}</p>
                  <p>Soil Health Index: ${document.getElementById('financial-gauge')?.querySelector('.label')?.textContent || 'N/A'}</p>
              </div>
              
              ${this.recommendations.map((rec, index) => `
                  <div class="recommendation">
                      <h3>${index + 1}. ${rec.title || 'Soil Improvement'}</h3>
                      ${rec.action ? `<p><strong>Action:</strong> ${rec.action}</p>` : ''}
                      ${rec.reason ? `<p><strong>Why:</strong> ${rec.reason}</p>` : ''}
                      ${rec.cost_estimate ? `<p><strong>Cost Estimate:</strong> ${rec.cost_estimate}</p>` : ''}
                      ${rec.timeframe ? `<p><strong>Timeframe:</strong> ${rec.timeframe}</p>` : ''}
                      ${rec.local_context ? `<p><strong>Local Context:</strong> ${rec.local_context}</p>` : ''}
                  </div>
              `).join('')}
              
              <div class="footer">
                  <p>Talazo AgriFinance Platform © ${new Date().getFullYear()}</p>
                  <p>Soil-based financial analytics for Zimbabwean farmers</p>
              </div>
          </body>
          </html>
      `;
      
      printWindow.document.write(html);
      printWindow.document.close();
      
      // Trigger print after content is loaded
      printWindow.addEventListener('load', () => {
          printWindow.print();
      });
  }
  
  shareRecommendations() {
      if (this.recommendations.length === 0) {
          if (window.dashboardManager) {
              window.dashboardManager.showNotification('No recommendations to share', 'warning');
          }
          return;
      }
      
      // Show sharing options (simplified for prototype)
      const options = ['Email', 'WhatsApp', 'SMS', 'Download PDF'];
      
      // Create a simple popup for sharing options
      const popup = document.createElement('div');
      popup.className = 'share-popup';
      popup.innerHTML = `
          <div class="share-popup-content">
              <h3>Share Recommendations</h3>
              <div class="share-options">
                  ${options.map(option => `
                      <button class="share-option" data-option="${option.toLowerCase()}">
                          <i class="fas fa-${option === 'Email' ? 'envelope' : option === 'WhatsApp' ? 'whatsapp' : option === 'SMS' ? 'sms' : 'file-pdf'}"></i>
                          ${option}
                      </button>
                  `).join('')}
              </div>
              <button class="close-popup"><i class="fas fa-times"></i></button>
          </div>
      `;
      
      document.body.appendChild(popup);
      
      // Add event listeners
      popup.querySelector('.close-popup').addEventListener('click', () => {
          popup.remove();
      });
      
      popup.querySelectorAll('.share-option').forEach(button => {
          button.addEventListener('click', () => {
              const option = button.dataset.option;
              popup.remove();
              
              if (window.dashboardManager) {
                  window.dashboardManager.showNotification(`Sharing via ${option} will be available in the next update`, 'default');
              }
          });
      });
  }
  
  getCostClass(costEstimate) {
      const cost = costEstimate.toLowerCase();
      if (cost.includes('low')) return 'cost-low';
      if (cost.includes('medium')) return 'cost-medium';
      if (cost.includes('high')) return 'cost-high';
      return '';
  }
  
  showLoading() {
      if (this.loadingElement) {
          this.loadingElement.classList.remove('hidden');
      }
      
      if (this.contentElement) {
          this.contentElement.classList.add('hidden');
      }
      
      if (this.generateButton) {
          this.generateButton.disabled = true;
      }
  }
  
  hideLoading() {
      if (this.loadingElement) {
          this.loadingElement.classList.add('hidden');
      }
      
      if (this.generateButton) {
          this.generateButton.disabled = false;
      }
  }
  
  showError(message) {
      if (this.contentElement) {
          this.contentElement.innerHTML = `
              <div class="error-message">
                  <i class="fas fa-exclamation-triangle"></i>
                  <p>${message}</p>
                  <p>Please try again or use our basic recommendations instead.</p>
              </div>
          `;
          this.contentElement.classList.remove('hidden');
      }
  }
}

// Initialize AI recommendations when DOM is loaded
document.addEventListener('DOMContentLoaded', () => {
  console.log('Initializing AI Recommendation Manager...');
  window.aiRecommendationManager = new AIRecommendationManager();
});